iothub/service/src/IoTHubSasCredential.cs

```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
```

Comment:
I had a chat with Scott on this and we landed on adding a derived class from AzureSasCredential. The reason we need this is that the AMQP library we use, takes the expiry date as input. Parsing the signature to find the expiry is not ideal and can cause bugs. So the best way forward for us is to create this derived class and take this as input in our constructors (in the upcoming PR).

Reply:
Does this just mean we are leaving the parsing logic to the user? Or do they have this date time field prior to creating the SAS token?

Reply:
From what I can tell, they should be constructing the SAS using a DateTime. For example: We do that when we get the connection string as input. Atleast for IotHub the portal only gives a connectionstring so I don't see any other way for them to have a connection string and parse it.

Reply:
The doc that I found (which I suspect customers will also reference) doesn't seem to give the users the expiration time for free:

```csharp
public static string generateSasToken(string resourceUri, string key, string policyName, int expiryInSeconds = 3600)
{
    TimeSpan fromEpochStart = DateTime.UtcNow - new DateTime(1970, 1, 1);
    string expiry = Convert.ToString((int)fromEpochStart.TotalSeconds + expiryInSeconds);

    string stringToSign = WebUtility.UrlEncode(resourceUri) + "\n" + expiry;

    HMACSHA256 hmac = new HMACSHA256(Convert.FromBase64String(key));
    string signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(stringToSign)));

    string token = String.Format(CultureInfo.InvariantCulture, "SharedAccessSignature sr={0}&sig={1}&se={2}", WebUtility.UrlEncode(resourceUri), WebUtility.UrlEncode(signature), expiry);

    if (!String.IsNullOrEmpty(policyName))
    {
        token += "&skn=" + policyName;
    }

    return token;
}
```

From [here](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security#security-token-structure)

Reply:
Good find! However I did not understand which way you were inclined for the design. The code snippet above shows that they have constructed a DateTime and appended to the SAS, which means they will have it handy to pass it on to the SDK. I feel parsing the SAS token again in the SDK might be redundant and could lead to bugs.

Reply:
I agree - avoid parsing

Reply:
@timtay-microsoft - Ping me if you want to discuss further. I will resolve this for now.